Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+              // to check generation.
+              if (!strong_this || strong_this->generation_ != auth_generation) {
+                return;
+              }
+              strong_this->ResumeStartAfterAuth(maybe_token);
+            });
+      });
+}
+
+void Stream::ResumeStartAfterAuth(const StatusOr<Token>& maybe_token) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Starting,
+              "State should still be 'Starting' (was %s)", state_);
+
+  if (!maybe_token.ok()) {
+    OnStreamError(maybe_token.status());
+    return;
+  }
+
+  absl::string_view token = [&] {
+    auto token = maybe_token.ValueOrDie();
+    return token.user().is_authenticated() ? token.token()
+                                           : absl::string_view{};
+  }();
+
+  grpc_stream_ = CreateGrpcStream(datastore_, token);
+  grpc_stream_->Start();
+}
+
+void Stream::OnStreamStart() {
+  EnsureOnQueue();
+
+  state_ = State::Open;
+  DoOnStreamStart();
+}
+
+// Backoff
+
+void Stream::BackoffAndTryRestarting() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s backoff", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Error,
+              "Should only perform backoff in an error case");
+
+  state_ = State::Backoff;
+  backoff_.BackoffAndRun([this] { ResumeStartFromBackoff(); });
+}
+
+void Stream::ResumeStartFromBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Backoff, "Backoff elapsed but state is now: %s",
+              state_);
+
+  state_ = State::Initial;
+  Start();
+  HARD_ASSERT(IsStarted(), "Stream should have started.");
+}
+
+void Stream::InhibitBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(!IsStarted(),
+              "Can only cancel backoff in a stopped state (was %s)", state_);
+
+  // Clear the error condition.
+  state_ = State::Initial;
+  ResetBackoff();
+}
+
+void Stream::ResetBackoff() {
+  EnsureOnQueue();
+  backoff_.Reset();
+}
+
+// Idleness
+
+void Stream::MarkIdle() {
+  EnsureOnQueue();
+
+  if (IsOpen() && !idleness_timer_) {
+    idleness_timer_ = worker_queue_->EnqueueAfterDelay(
+        kIdleTimeout, idle_timer_id_, [this] { Stop(); });
+  }
+}
+
+void Stream::CancelIdleCheck() {
+  EnsureOnQueue();
+  idleness_timer_.Cancel();
+}
+
+// Read/write
+
+void Stream::OnStreamRead(const grpc::ByteBuffer& message) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(IsStarted(), "OnStreamRead called for a stopped stream.");
+
+  if (bridge::IsLoggingEnabled()) {
+    LOG_DEBUG("%s headers (whitelisted): %s", GetDebugDescription(),
+              Datastore::GetWhitelistedHeadersAsString(
+                  grpc_stream_->GetResponseHeaders()));
+  }
+
+  Status read_status = DoOnStreamRead(message);
+  if (!read_status.ok()) {
+    grpc_stream_->Finish();
+    // Don't expect GRPC to produce status -- since the error happened on the
+    // client, we have all the information we need.
+    OnStreamError(read_status);
+    return;
+  }
+}
+
+// Stopping
+
+void Stream::Stop() {
```

Comment:
I'm not a fan of this refactoring because there's so much that these methods have to do that's common and it's difficult to tell if they're both doing it correctly.

Closing the stream requires us to do the following:
  * Cancel timers (common)
  * Bump close count (common)
  * Handle stream status (slightly different action in both cases but as a step it's common)
  * Give the subclass a chance to flush the stream token (success only)
  * Force close the stream (common)
  * Assign final state (common)
  * Notify the delegate (common)

A way to bring the apparent complexity of this method down would be to factor out "handle stream status" into a separate method.

In this case, neither is doing this exactly right.

Reply:
Thanks for the write-up; I can now see where you're coming from, and it looks like I underestimated the risk of divergence. Changed the function to be a lot more in line with existing implementations. Here's a quick list of intentional differences:
* Added a comment with the rationale and explicitly marked the steps.
* Moved the `IsStarted` check into the method.
* Moved some of the verbose status handling into a helper function.
* Probably most notable change -- I removed the passing of `final_state` as an argument. Instead, `final_state` is calculated within the function based on whether the status is ok. Whether this is an improvement depends a lot on whether you think there are valid use cases for final states other than `Initial` and `Error` (there aren't currently, but perhaps it's better to not set this in stone). Otherwise, I would still prefer an explicit assertion on the final state, to make sure nonsensical states cannot be passed to this function.

I'd like to stress that I don't feel strongly about any of these changes -- feel free to push back on any/all of them.

Reply:
I don't think we had an explicit plan for final states other than Initial or Error--when going through the refactoring to create a single method in the wake of the latest report of an assertion failure it wasn't clear at the time if we would need it or not. I'm happy to clean this up now and revisit if we ever do actually need that kind of flexibility. I doubt we will.